### PR TITLE
Adding support for getting the identifier from the id_token

### PR DESCRIPTION
**Pull request by stevensajw**
This PR: 
- adds support to fetch the user identifier from the idtoken instead of the userinfo endpoint.
- adds configuration parameter for controlling it with `user_identifier_from_idtoken`
- adds documentation for the feature 
- introduces a new `OidcJwtHelper->getIdTokenClaims()` method getIdTokenClaims() and use it in `verifyTokens`.

Not sure about: 
I use OidcClient to get the data from the token as it already had the token helper. 

For issue #37

### CODE REVIEW

File: src/OidcClient.php [hunk 1]
```diff
@@ -218,6 +218,11 @@ public function retrieveUserInfo(OidcTokens $tokens): OidcUserData
     return new OidcUserData($data);
   }
 
+ /** {@inheritDoc} */
+  public function retrieveDataFromIdToken(OidcTokens $tokens): array {
+      return json_decode(json_encode($this->jwtHelper->getIdTokenClaims($tokens)), true);
```
Comment (bobvandevijver) [inline src/OidcClient.php:223]: Why did you add this encode-decode step?  Is this purely to convert the object from `getIdTokenClaims` into an array? A cast (`(array)`) would be sufficient, but retrieving the property with either the property accessor or just directly with `$data->{$property}` would also be valid options.

File: src/OidcClientInterface.php [hunk 2]
```diff
@@ -75,4 +75,12 @@ public function generateEndSessionEndpointRedirect(
    * @throws OidcException
    */
   public function retrieveUserInfo(OidcTokens $tokens): OidcUserData;
+
+  /**
+   * Retrieve additional information from the idtoken.
+   *
+   * @throws OidcException
+   */
+  public function retrieveDataFromIdToken(OidcTokens $tokens): array;
```
Comment (bobvandevijver) [inline src/OidcClientInterface.php:84]: Adding a method to the interface would be a BC break for people implementing this interface, so I am a bit hesitant about this.

File: src/Security/OidcAuthenticator.php [hunk 3]
```diff
@@ -63,8 +64,18 @@ public function authenticate(Request $request): Passport
...
 
+      // Look for the user identifier in either the id_token or the userinfo endpoint
+      if ($this->userIdentifierFromIdToken) {
+        $tokenData = $this->oidcClient->retrieveDataFromIdToken($authData);
+        $userIdentifier = array_key_exists($this->userIdentifierProperty, $tokenData)
+            ? $tokenData[$this->userIdentifierProperty]
+            : null;
+      } else {
+        $userIdentifier = $userData->getUserDataString($this->userIdentifierProperty);
+      }
```
Comment (bobvandevijver) [inline src/Security/OidcAuthenticator.php:75]: With this code I would prefer the tertiary assignment.

```php
      $userIdentifier = $this->userIdentifierFromIdToken
        ? $this->oidcClient->retrieveDataFromIdToken($authData)[$this->userIdentifierProperty] ?? null
        : $userData->getUserDataString($this->userIdentifierProperty);
```

Comment (bobvandevijver) [reply src/Security/OidcAuthenticator.php:75]: But (combined with the other comments actually), lets use the jwt helper directly here, something like: 

```php
$this->jwtHelper->getIdTokenClaims($authData)->{$this->userIdentifierProperty} ?? null;
```

The JwtHelper will need to be injected in the authenticator as well, but it can be an optional dependency with null default. That way we do not break BC, but we should throw when it is null and the `userIdentifierFromIdToken` is set to `true`. So then it becomes something like this:

```php
if ($this->userIdentifierFromIdToken) {
  if (!$this->jwtHelper) {
    throw new RuntimeException('Missing required JwtHelper object when retrieving user identifier from token. Did you extend the authenticator and not provide the JwtHelper?')
  }

  $userIdentifier = $this->jwtHelper->getIdTokenClaims($authData)->{$this->userIdentifierProperty} ?? null;
} else {
  $userIdentifier = $userData->getUserDataString($this->userIdentifierProperty);
}
```

File: src/Security/OidcAuthenticator.php [hunk 4]
```diff
@@ -37,6 +37,7 @@ public function __construct(
       private string $checkPath,
       private string $loginPath,
       private string $userIdentifierProperty,
+      private bool $userIdentifierFromIdToken,
       private bool $enableRememberMe)
```
Comment (bobvandevijver) [inline src/Security/OidcAuthenticator.php:41]:
```suggestion
      private bool $enableRememberMe,
      private bool $userIdentifierFromIdToken = false)
```

For BC reasons.

File: src/Security/Factory/OidcFactory.php [hunk 5]
```diff
@@ -60,6 +61,7 @@ public function createAuthenticator(
         ->addArgument($config['check_path'])
         ->addArgument($config['login_path'])
         ->addArgument($config['user_identifier_property'])
+        ->addArgument($config['user_identifier_from_idtoken'])
         ->addArgument($config['enable_remember_me']);
```
Comment (bobvandevijver) [inline src/Security/Factory/OidcFactory.php:65]:
```suggestion
        ->addArgument($config['enable_remember_me'])
        ->addArgument($config['user_identifier_from_idtoken']);
```

For BC reasons of the class.

**Review by bobvandevijver — CHANGES_REQUESTED** [review]
Love the PR, but I do have a couple of questions/remarks which are mostly related to ensure BC when merged!

**Comment by stevensajw** [comment]
I updated the commit. Thank you very much for the feedback. I never worked on a Symfony bundle before. I was unsure how to use the injection in the bundle configuration, so I tried working around it, creating some new issues along the way. It looks better like this. 

Thanks
AJ

**Review by bobvandevijver — APPROVED** [review]
